### ISSUE TITLE
Adding a new Feature

### ISSUE THREAD

**Issue by rohitrj2ee** [togglz/togglz#234]
Is there an easy way to add/maintain new features? I am currently implementing by the file based approach. Let's say i have to add a new feature. Can i add a new feature through the console without updating the features enum file and the properties file?

**Comment by chkal** [comment]
No, the console doesn't support adding new feature at runtime. Especially using the Java enum based approach will of course not work if you want to add features at runtime. You could however implement a custom `FeatureProvider` to achieve something like that.

However, in my opinion adding features at runtime typically doesn't make much sense because your code has to know about and actively check the features to select different branches to execute. Or are you trying to achieve something else?

**Comment by amit-lasry** [comment]
Pardon for jumping into this discussion, but I wanted to share that we're facing the same dilemma.

Ideally it would be awesome to have a single feature-control system across our eco-system (web,mobile,backend...).
Following what @rohitrj2ee wrote, it only makes sense to add a feature in runtime if it's a **client** (mobile/web) feature and the backend is simply channelling the feature state to the clients.

IMHO that's a viable scenario that would be very helpful if addressed.

BTW - A "similar" (but less rich) example exists in Google Firebase (a remote configuration mgmt. tool) that allows adding key/value pairs and updating the clients in runtime, which enables them to change behaviour regardless of the backend implementation.

**Comment by chkal** [comment]
Hey @amit-lasry,

thanks for sharing your thoughts on this. Back then, the idea of Togglz was to create a feature toggle solution for single applications. Therefore, the enum approach totally made sense. But I agree that if you want to distribute feature flags across your infrastructure, the enum approach doesn't work very well. Although I'm not 100% convinced if it is a good idea to distribute such toggles. At least for the server side parts. But that's a different story.

However, the `FeatureProvider` contract should actually provide you everything you need. If not, please raise your voice. You will actually have to write the UI part yourself, because the admin console assumes static feature metadata, but that should be fairly simply. But of course Toggle could also provide more `FeatureProvider` implementations out of the box. Just let us know about your ideas. And PRs are welcome. :+1:

**Comment by joewarner** [comment]
Hi @chkal,

I'm still finding my way around Togglz and need to look deeper into the custom `FeatureProvider` that you suggest but wanted to add my use case here too for completeness.

In my case, although I can largely treat my application as standalone, it is deployed to an "application server" that - to all intents and purposes - is permanently running. Effectively my application is hot-deployed each time it changes, without a need for restarting the container. (In fact the owner of the container actively discourages restarting just because applications are being re-deployed.) I completely understand and agree that "adding features at runtime typically doesn't make much sense" to a fully standalone application and whilst I accept that my scenario is somewhat unusual it does seem that there are others out there with similar issues. 

Thus the two key features that I am looking for are

1. The ability to reload features from the repository (either file or database)
1. Add new features effectively on-the-fly by extending the repository and reloading it

As you point out the `enum` approach doesn't seem well suited to either of these requirements. My challenge now is whether to continue to pursue the `FeatureProvider` option or whether it would make more sense for me to look for a different library (I'm unlikely to develop my own implementation at this stage).

**Comment by chkal** [comment]
@joewarner I'm not sure if I fully understand you case correctly. So you are deploying your application to an application server. That's fine. Your feature enum is included in your application, correct? In this case even hot deployments should work fine and you should get the new enum with each deployment. Assuming a traditional application server where each deployment results in a new class loader.

**Comment by joewarner** [comment]
I tried to indicate that it isn't actually an app server but rather an "application server", i.e. something that fulfills that role. (If you are really interested then I work with the IBM BPM product. So, the code that I deploy runs inside an application (BPM) that is itself running on the app server and the code I deploy isn't even really Java!)

But yes, you are right, the enum is included in the application. And I can see what you say about hot deployments should work in that scenario. However, in that environment it is a bit of an unnatural process to have to recompile this little bit of Java each time you want to add a new Feature Flag. I do understand why it is necessary but I think I am increasingly of the opinion that `enum`s are simply going to be too cumbersome for me in this specific scenario.

However, I do still need to understand how, once that application is deployed to Production I can change the values of the Toggles without doing a restart - unfortunately I haven't yet had an opportunity to investigate the `FeatureProvider` option you previously mentioned.

I have since been looking at FF4j. Its API certainly seems to be better suited to what I think I want to achieve. However, it is also very complex (one might say bloated) for what I had thought should be a fairly simple problem. So, I haven't discounted Togglz just yet, but I am considering other options.

**Comment by chkal** [comment]
Thanks for sharing these details.

I think a custom `FeatureProvider` could help in this case. The API is really simple:

https://github.com/togglz/togglz/blob/master/core/src/main/java/org/togglz/core/spi/FeatureProvider.java

Basically a `FeatureProvider` does the following:
* It tells Togglz about all the features your application supports. In your case the result of this could be dynamic. The default `EnumFeatureProvider` just builds the list from the enum
* It can tell Togglz about feature metadata (label, default value, etc). The default `EnumFeatureProvider` builds this metadata from annotations.

There is already a `FeatureProvider` implementation based on Properties. Not sure if this one can help you in your situation, but it may be a good start to learn how to create your own `FeatureProvider`.

https://github.com/togglz/togglz/blob/master/core/src/main/java/org/togglz/core/manager/PropertyFeatureProvider.java

**Comment by guidosteinacker1969** [comment]
Everybody happy with closing this issue?

**Comment by cranberrysoft** [comment]
In my case I need to run admin panel on a separate instance then my web-application that uses features. First reason is security and the second reason is business requirement - we want to have feature flags integrated with current admin panel.
 
This means that if I want to add a new feature flag I would need to release two applications: one that contains admin panel and one that uses features which is not the most convenient way.

In this case the dynamic behaviour would be desired.

**Comment by hennr** [comment]
This is out of scope for now, sorry.
